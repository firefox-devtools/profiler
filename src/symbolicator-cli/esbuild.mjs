import esbuild from 'esbuild';
import fs from 'fs';
import { nodeBaseConfig } from '../../esbuild.mjs';

async function buildAll() {
  // Clean dist directory
  if (fs.existsSync('dist')) {
    fs.rmSync('dist', { recursive: true });
  }

  await esbuild.build({
    ...nodeBaseConfig,
    entryPoints: ['src/symbolicator-cli/index.ts'],
    outfile: 'dist/symbolicator-cli.js',
  });

  console.log('âœ… Build completed');
}

// Run build if called directly
if (import.meta.url === `file://${process.argv[1]}`) {
  buildAll().catch(console.error);
}
