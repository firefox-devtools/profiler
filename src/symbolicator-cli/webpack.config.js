const path = require('path');
const projectRoot = path.join(__dirname, '../..');
const includes = [path.join(projectRoot, 'src')];

module.exports = {
  name: 'symbolicator-cli',
  target: 'node',
  mode: process.env.NODE_ENV,
  resolve: {
    extensions: ['.js', '.jsx', '.ts', '.tsx'],
  },
  output: {
    path: path.resolve(projectRoot, 'dist'),
    filename: 'symbolicator-cli.js',
  },
  entry: './src/symbolicator-cli/index.ts',
  module: {
    rules: [
      {
        test: /\.(js|ts|tsx)$/,
        use: ['babel-loader'],
        include: includes,
      },
      {
        test: /\.svg$/,
        type: 'asset/resource',
      },
    ],
  },
  experiments: {
    // Make WebAssembly work just like in webpack v4
    syncWebAssembly: true,
  },
};
