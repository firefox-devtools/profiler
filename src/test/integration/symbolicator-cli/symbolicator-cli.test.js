/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import fs from 'fs';
import os from 'os';
import path from 'path';
import { run } from '../../../symbolicator-cli';

describe('symbolicator-cli tool', function () {
  async function runToTempFileAndReturnOutput(options) {
    const tempDir = fs.mkdtempSync(
      path.join(os.tmpdir(), 'symbolicator-cli-test')
    );
    const tempFile = path.join(tempDir, 'temp.json');
    options.output = tempFile;

    try {
      await run(options);
      return JSON.parse(fs.readFileSync(tempFile, 'utf-8'));
    } finally {
      // $FlowExpectError Flow doesn't know about the rmSync API despite it's been implemented in node v16. Sigh
      fs.rmSync(tempDir, { recursive: true, force: true });
    }
  }

  it('is symbolicating a trace correctly', async function () {
    jest.spyOn(console, 'log').mockImplementation(() => {});
    jest.spyOn(console, 'warn').mockImplementation(() => {});

    const symbolsJson = fs.readFileSync(
      'src/test/integration/symbolicator-cli/symbol-server-response.json'
    );

    window.fetchMock.post(
      'http://symbol.server/symbolicate/v5',
      new Response(symbolsJson)
    );

    const options = {
      input: 'src/test/integration/symbolicator-cli/unsymbolicated.json',
      output: '',
      server: 'http://symbol.server',
    };

    const result = await runToTempFileAndReturnOutput(options);

    expect(console.warn).not.toHaveBeenCalled();
    expect(result).toMatchSnapshot();
  });
});
