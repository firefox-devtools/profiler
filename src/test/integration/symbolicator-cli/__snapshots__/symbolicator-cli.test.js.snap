// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`symbolicator-cli tool is symbolicating a trace correctly 1`] = `
Object {
  "counters": Array [],
  "libs": Array [
    Object {
      "arch": "arm64e",
      "breakpadId": "F635824E318B3F0C842CC369737F2B680",
      "codeId": "F635824E318B3F0C842CC369737F2B68",
      "debugName": "dyld",
      "debugPath": "/usr/lib/dyld",
      "name": "dyld",
      "path": "/usr/lib/dyld",
    },
    Object {
      "arch": "arm64",
      "breakpadId": "F61DA4D57CBB38CA8BDF059C645834520",
      "codeId": "F61DA4D57CBB38CA8BDF059C64583452",
      "debugName": "a.out",
      "debugPath": "/usr/helloworld/a.out",
      "name": "a.out",
      "path": "/usr/helloworld/a.out",
    },
    Object {
      "arch": "arm64e",
      "breakpadId": "E03E84786F5C3D21A79A58408F5140000",
      "codeId": "E03E84786F5C3D21A79A58408F514000",
      "debugName": "libsystem_pthread.dylib",
      "debugPath": "/usr/lib/system/libsystem_pthread.dylib",
      "name": "libsystem_pthread.dylib",
      "path": "/usr/lib/system/libsystem_pthread.dylib",
    },
    Object {
      "arch": "arm64e",
      "breakpadId": "71FF45B8F14E36669E966CF58315B91D0",
      "codeId": "71FF45B8F14E36669E966CF58315B91D",
      "debugName": "libsystem_kernel.dylib",
      "debugPath": "/usr/lib/system/libsystem_kernel.dylib",
      "name": "libsystem_kernel.dylib",
      "path": "/usr/lib/system/libsystem_kernel.dylib",
    },
    Object {
      "arch": "arm64e",
      "breakpadId": "D30F183093D03D0B8CBA9544E84BFD5B0",
      "codeId": "D30F183093D03D0B8CBA9544E84BFD5B",
      "debugName": "libsystem_c.dylib",
      "debugPath": "/usr/lib/system/libsystem_c.dylib",
      "name": "libsystem_c.dylib",
      "path": "/usr/lib/system/libsystem_c.dylib",
    },
    Object {
      "arch": "arm64e",
      "breakpadId": "B4BF9F8931D737428CE7AB3554F9F5250",
      "codeId": "B4BF9F8931D737428CE7AB3554F9F525",
      "debugName": "libsystem_platform.dylib",
      "debugPath": "/usr/lib/system/libsystem_platform.dylib",
      "name": "libsystem_platform.dylib",
      "path": "/usr/lib/system/libsystem_platform.dylib",
    },
  ],
  "meta": Object {
    "categories": Array [
      Object {
        "color": "grey",
        "name": "Other",
        "subcategories": Array [
          "Other",
        ],
      },
      Object {
        "color": "yellow",
        "name": "User",
        "subcategories": Array [
          "Other",
        ],
      },
    ],
    "debug": false,
    "extensions": Object {
      "baseURL": Array [],
      "id": Array [],
      "length": 0,
      "name": Array [],
    },
    "interval": 1,
    "markerSchema": Array [],
    "oscpu": "macOS 14.6.1",
    "pausedRanges": Array [],
    "preprocessedProfileVersion": 55,
    "processType": 0,
    "product": "a.out",
    "sampleUnits": Object {
      "eventDelay": "ms",
      "threadCPUDelta": "Âµs",
      "time": "ms",
    },
    "sourceCodeIsNotOnSearchfox": true,
    "startTime": 1726433495880.2869,
    "symbolicated": true,
    "usesOnlyOneStackType": true,
    "version": 24,
  },
  "pages": Array [],
  "profilerOverhead": Array [],
  "threads": Array [
    Object {
      "frameTable": Object {
        "address": Array [
          24915,
          16067,
          38027,
          11180,
        ],
        "category": Array [
          1,
          1,
          1,
          1,
        ],
        "column": Array [
          null,
          null,
          null,
          null,
        ],
        "func": Array [
          0,
          1,
          2,
          3,
        ],
        "inlineDepth": Array [
          0,
          0,
          0,
          0,
        ],
        "innerWindowID": Array [
          null,
          null,
          null,
          null,
        ],
        "length": 4,
        "line": Array [
          null,
          null,
          null,
          null,
        ],
        "nativeSymbol": Array [
          0,
          1,
          2,
          3,
        ],
        "subcategory": Array [
          0,
          0,
          0,
          0,
        ],
      },
      "funcTable": Object {
        "columnNumber": Array [
          null,
          null,
          null,
          null,
        ],
        "fileName": Array [
          null,
          null,
          null,
          null,
        ],
        "isJS": Array [
          false,
          false,
          false,
          false,
        ],
        "length": 4,
        "lineNumber": Array [
          null,
          null,
          null,
          null,
        ],
        "name": Array [
          8,
          9,
          10,
          11,
        ],
        "relevantForJS": Array [
          false,
          false,
          false,
          false,
        ],
        "resource": Array [
          0,
          1,
          2,
          3,
        ],
      },
      "isMainThread": true,
      "markers": Object {
        "category": Array [],
        "data": Array [],
        "endTime": Array [],
        "length": 0,
        "name": Array [],
        "phase": Array [],
        "startTime": Array [],
      },
      "name": "a.out",
      "nativeSymbols": Object {
        "address": Array [
          22440,
          15828,
          37420,
          11172,
        ],
        "functionSize": Array [
          2788,
          380,
          1028,
          44,
        ],
        "length": 4,
        "libIndex": Array [
          0,
          1,
          2,
          3,
        ],
        "name": Array [
          8,
          9,
          10,
          11,
        ],
      },
      "pausedRanges": Array [],
      "pid": "56127",
      "processName": "a.out",
      "processShutdownTime": 300.814417,
      "processStartupTime": 237.805292,
      "processType": "default",
      "registerTime": 237.805292,
      "resourceTable": Object {
        "host": Array [
          null,
          null,
          null,
          null,
        ],
        "length": 4,
        "lib": Array [
          0,
          1,
          2,
          3,
        ],
        "name": Array [
          0,
          2,
          4,
          6,
        ],
        "type": Array [
          1,
          1,
          1,
          1,
        ],
      },
      "samples": Object {
        "length": 11,
        "stack": Array [
          3,
          3,
          3,
          3,
          3,
          3,
          3,
          3,
          3,
          3,
          3,
        ],
        "threadCPUDelta": Array [
          11273,
          0,
          8,
          0,
          3,
          4,
          0,
          30,
          0,
          21,
          0,
        ],
        "time": Array [
          240.520375,
          251.722709,
          252.728334,
          261.724667,
          262.732625,
          263.723459,
          274.768584,
          275.7785,
          286.765084,
          287.756375,
          299.7795,
        ],
        "weight": Array [
          1,
          11,
          1,
          9,
          1,
          1,
          11,
          1,
          11,
          1,
          12,
        ],
        "weightType": "samples",
      },
      "stackTable": Object {
        "frame": Array [
          0,
          1,
          2,
          3,
        ],
        "length": 4,
        "prefix": Array [
          null,
          0,
          1,
          2,
        ],
      },
      "stringArray": Array [
        "dyld",
        "0x6153",
        "a.out",
        "0x3ec3",
        "libsystem_pthread.dylib",
        "0x948b",
        "libsystem_kernel.dylib",
        "0x2bac",
        "start",
        "main",
        "_pthread_join",
        "__ulock_wait",
      ],
      "tid": "6274156",
      "unregisterTime": 300.814417,
    },
    Object {
      "frameTable": Object {
        "address": Array [
          28563,
          15795,
          15667,
          54399,
          54631,
          17384,
        ],
        "category": Array [
          1,
          1,
          1,
          1,
          1,
          1,
        ],
        "column": Array [
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "func": Array [
          0,
          1,
          2,
          3,
          4,
          5,
        ],
        "inlineDepth": Array [
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        "innerWindowID": Array [
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "length": 6,
        "line": Array [
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "nativeSymbol": Array [
          2,
          0,
          1,
          4,
          5,
          3,
        ],
        "subcategory": Array [
          0,
          0,
          0,
          0,
          0,
          0,
        ],
      },
      "funcTable": Object {
        "columnNumber": Array [
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "fileName": Array [
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "isJS": Array [
          false,
          false,
          false,
          false,
          false,
          false,
        ],
        "length": 6,
        "lineNumber": Array [
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "name": Array [
          12,
          10,
          11,
          14,
          15,
          13,
        ],
        "relevantForJS": Array [
          false,
          false,
          false,
          false,
          false,
          false,
        ],
        "resource": Array [
          0,
          1,
          1,
          2,
          2,
          3,
        ],
      },
      "isMainThread": false,
      "markers": Object {
        "category": Array [],
        "data": Array [],
        "endTime": Array [],
        "length": 0,
        "name": Array [],
        "phase": Array [],
        "startTime": Array [],
      },
      "name": "Thread <6274161>",
      "nativeSymbols": Object {
        "address": Array [
          15752,
          15644,
          28428,
          17376,
          54332,
          54412,
        ],
        "functionSize": Array [
          76,
          108,
          320,
          44,
          80,
          464,
        ],
        "length": 6,
        "libIndex": Array [
          1,
          1,
          2,
          3,
          4,
          4,
        ],
        "name": Array [
          10,
          11,
          12,
          13,
          14,
          15,
        ],
      },
      "pausedRanges": Array [],
      "pid": "56127",
      "processName": "a.out",
      "processShutdownTime": 300.814417,
      "processStartupTime": 237.805292,
      "processType": "default",
      "registerTime": 240.520375,
      "resourceTable": Object {
        "host": Array [
          null,
          null,
          null,
          null,
        ],
        "length": 4,
        "lib": Array [
          2,
          1,
          4,
          3,
        ],
        "name": Array [
          0,
          2,
          5,
          8,
        ],
        "type": Array [
          1,
          1,
          1,
          1,
        ],
      },
      "samples": Object {
        "length": 2,
        "stack": Array [
          5,
          5,
        ],
        "threadCPUDelta": Array [
          7,
          0,
        ],
        "time": Array [
          240.520375,
          251.722709,
        ],
        "weight": Array [
          1,
          11,
        ],
        "weightType": "samples",
      },
      "stackTable": Object {
        "frame": Array [
          0,
          1,
          2,
          3,
          4,
          5,
        ],
        "length": 6,
        "prefix": Array [
          null,
          0,
          1,
          2,
          3,
          4,
        ],
      },
      "stringArray": Array [
        "libsystem_pthread.dylib",
        "0x6f93",
        "a.out",
        "0x3db3",
        "0x3d33",
        "libsystem_c.dylib",
        "0xd47f",
        "0xd567",
        "libsystem_kernel.dylib",
        "0x43e8",
        "threadfunc(void*)",
        "fac(unsigned long)",
        "_pthread_start",
        "__semwait_signal",
        "usleep",
        "nanosleep",
      ],
      "tid": "6274161",
      "unregisterTime": 252.728334,
    },
    Object {
      "frameTable": Object {
        "address": Array [
          28563,
          15795,
          15667,
          54399,
          54631,
          17384,
          15719,
        ],
        "category": Array [
          1,
          1,
          1,
          1,
          1,
          1,
          1,
        ],
        "column": Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "func": Array [
          0,
          1,
          2,
          3,
          4,
          5,
          2,
        ],
        "inlineDepth": Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        "innerWindowID": Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "length": 7,
        "line": Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "nativeSymbol": Array [
          2,
          0,
          1,
          4,
          5,
          3,
          1,
        ],
        "subcategory": Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
      },
      "funcTable": Object {
        "columnNumber": Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "fileName": Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "isJS": Array [
          false,
          false,
          false,
          false,
          false,
          false,
          false,
        ],
        "length": 7,
        "lineNumber": Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "name": Array [
          13,
          11,
          12,
          15,
          16,
          14,
          10,
        ],
        "relevantForJS": Array [
          false,
          false,
          false,
          false,
          false,
          false,
          false,
        ],
        "resource": Array [
          0,
          1,
          1,
          2,
          2,
          3,
          1,
        ],
      },
      "isMainThread": false,
      "markers": Object {
        "category": Array [],
        "data": Array [],
        "endTime": Array [],
        "length": 0,
        "name": Array [],
        "phase": Array [],
        "startTime": Array [],
      },
      "name": "Thread <6274162>",
      "nativeSymbols": Object {
        "address": Array [
          15752,
          15644,
          28428,
          17376,
          54332,
          54412,
        ],
        "functionSize": Array [
          76,
          108,
          320,
          44,
          80,
          464,
        ],
        "length": 6,
        "libIndex": Array [
          1,
          1,
          2,
          3,
          4,
          4,
        ],
        "name": Array [
          11,
          12,
          13,
          14,
          15,
          16,
        ],
      },
      "pausedRanges": Array [],
      "pid": "56127",
      "processName": "a.out",
      "processShutdownTime": 300.814417,
      "processStartupTime": 237.805292,
      "processType": "default",
      "registerTime": 240.520375,
      "resourceTable": Object {
        "host": Array [
          null,
          null,
          null,
          null,
        ],
        "length": 4,
        "lib": Array [
          2,
          1,
          4,
          3,
        ],
        "name": Array [
          0,
          2,
          5,
          8,
        ],
        "type": Array [
          1,
          1,
          1,
          1,
        ],
      },
      "samples": Object {
        "length": 4,
        "stack": Array [
          5,
          5,
          10,
          10,
        ],
        "threadCPUDelta": Array [
          3,
          0,
          5,
          0,
        ],
        "time": Array [
          240.520375,
          251.722709,
          252.728334,
          261.724667,
        ],
        "weight": Array [
          1,
          11,
          1,
          9,
        ],
        "weightType": "samples",
      },
      "stackTable": Object {
        "frame": Array [
          0,
          1,
          2,
          3,
          4,
          5,
          6,
          2,
          3,
          4,
          5,
        ],
        "length": 11,
        "prefix": Array [
          null,
          0,
          1,
          2,
          3,
          4,
          1,
          6,
          7,
          8,
          9,
        ],
      },
      "stringArray": Array [
        "libsystem_pthread.dylib",
        "0x6f93",
        "a.out",
        "0x3db3",
        "0x3d33",
        "libsystem_c.dylib",
        "0xd47f",
        "0xd567",
        "libsystem_kernel.dylib",
        "0x43e8",
        "0x3d67",
        "threadfunc(void*)",
        "fac(unsigned long)",
        "_pthread_start",
        "__semwait_signal",
        "usleep",
        "nanosleep",
      ],
      "tid": "6274162",
      "unregisterTime": 262.732625,
    },
    Object {
      "frameTable": Object {
        "address": Array [
          28563,
          15795,
          15667,
          54399,
          54631,
          17384,
          15719,
        ],
        "category": Array [
          1,
          1,
          1,
          1,
          1,
          1,
          1,
        ],
        "column": Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "func": Array [
          0,
          1,
          2,
          3,
          4,
          5,
          2,
        ],
        "inlineDepth": Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        "innerWindowID": Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "length": 7,
        "line": Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "nativeSymbol": Array [
          2,
          0,
          1,
          4,
          5,
          3,
          1,
        ],
        "subcategory": Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
      },
      "funcTable": Object {
        "columnNumber": Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "fileName": Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "isJS": Array [
          false,
          false,
          false,
          false,
          false,
          false,
          false,
        ],
        "length": 7,
        "lineNumber": Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "name": Array [
          13,
          11,
          12,
          15,
          16,
          14,
          10,
        ],
        "relevantForJS": Array [
          false,
          false,
          false,
          false,
          false,
          false,
          false,
        ],
        "resource": Array [
          0,
          1,
          1,
          2,
          2,
          3,
          1,
        ],
      },
      "isMainThread": false,
      "markers": Object {
        "category": Array [],
        "data": Array [],
        "endTime": Array [],
        "length": 0,
        "name": Array [],
        "phase": Array [],
        "startTime": Array [],
      },
      "name": "Thread <6274163>",
      "nativeSymbols": Object {
        "address": Array [
          15752,
          15644,
          28428,
          17376,
          54332,
          54412,
        ],
        "functionSize": Array [
          76,
          108,
          320,
          44,
          80,
          464,
        ],
        "length": 6,
        "libIndex": Array [
          1,
          1,
          2,
          3,
          4,
          4,
        ],
        "name": Array [
          11,
          12,
          13,
          14,
          15,
          16,
        ],
      },
      "pausedRanges": Array [],
      "pid": "56127",
      "processName": "a.out",
      "processShutdownTime": 300.814417,
      "processStartupTime": 237.805292,
      "processType": "default",
      "registerTime": 240.520375,
      "resourceTable": Object {
        "host": Array [
          null,
          null,
          null,
          null,
        ],
        "length": 4,
        "lib": Array [
          2,
          1,
          4,
          3,
        ],
        "name": Array [
          0,
          2,
          5,
          8,
        ],
        "type": Array [
          1,
          1,
          1,
          1,
        ],
      },
      "samples": Object {
        "length": 6,
        "stack": Array [
          5,
          5,
          10,
          10,
          15,
          15,
        ],
        "threadCPUDelta": Array [
          1,
          0,
          5,
          0,
          3,
          0,
        ],
        "time": Array [
          240.520375,
          251.722709,
          252.728334,
          263.723459,
          264.733042,
          274.768584,
        ],
        "weight": Array [
          1,
          11,
          1,
          11,
          1,
          10,
        ],
        "weightType": "samples",
      },
      "stackTable": Object {
        "frame": Array [
          0,
          1,
          2,
          3,
          4,
          5,
          6,
          2,
          3,
          4,
          5,
          6,
          2,
          3,
          4,
          5,
        ],
        "length": 16,
        "prefix": Array [
          null,
          0,
          1,
          2,
          3,
          4,
          1,
          6,
          7,
          8,
          9,
          6,
          11,
          12,
          13,
          14,
        ],
      },
      "stringArray": Array [
        "libsystem_pthread.dylib",
        "0x6f93",
        "a.out",
        "0x3db3",
        "0x3d33",
        "libsystem_c.dylib",
        "0xd47f",
        "0xd567",
        "libsystem_kernel.dylib",
        "0x43e8",
        "0x3d67",
        "threadfunc(void*)",
        "fac(unsigned long)",
        "_pthread_start",
        "__semwait_signal",
        "usleep",
        "nanosleep",
      ],
      "tid": "6274163",
      "unregisterTime": 275.7785,
    },
    Object {
      "frameTable": Object {
        "address": Array [
          28563,
          15795,
          15667,
          54399,
          54631,
          17384,
          15719,
          28575,
          30367,
          18819,
          16188,
        ],
        "category": Array [
          1,
          1,
          1,
          1,
          1,
          1,
          1,
          1,
          1,
          1,
          1,
        ],
        "column": Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "func": Array [
          0,
          1,
          2,
          3,
          4,
          5,
          2,
          0,
          7,
          8,
          10,
        ],
        "inlineDepth": Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        "innerWindowID": Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "length": 11,
        "line": Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "nativeSymbol": Array [
          2,
          0,
          1,
          6,
          7,
          5,
          1,
          2,
          3,
          4,
          8,
        ],
        "subcategory": Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
      },
      "funcTable": Object {
        "columnNumber": Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "fileName": Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "isJS": Array [
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
        ],
        "length": 11,
        "lineNumber": Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "name": Array [
          18,
          16,
          17,
          22,
          23,
          21,
          10,
          19,
          20,
          13,
          24,
        ],
        "relevantForJS": Array [
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
        ],
        "resource": Array [
          0,
          1,
          1,
          2,
          2,
          3,
          1,
          0,
          0,
          0,
          4,
        ],
      },
      "isMainThread": false,
      "markers": Object {
        "category": Array [],
        "data": Array [],
        "endTime": Array [],
        "length": 0,
        "name": Array [],
        "phase": Array [],
        "startTime": Array [],
      },
      "name": "Thread <6274164>",
      "nativeSymbols": Object {
        "address": Array [
          15752,
          15644,
          28428,
          30256,
          18740,
          17376,
          54332,
          54412,
          16080,
        ],
        "functionSize": Array [
          76,
          108,
          320,
          120,
          92,
          44,
          80,
          464,
          212,
        ],
        "length": 9,
        "libIndex": Array [
          1,
          1,
          2,
          2,
          2,
          3,
          4,
          4,
          5,
        ],
        "name": Array [
          16,
          17,
          18,
          19,
          20,
          21,
          22,
          23,
          24,
        ],
      },
      "pausedRanges": Array [],
      "pid": "56127",
      "processName": "a.out",
      "processShutdownTime": 300.814417,
      "processStartupTime": 237.805292,
      "processType": "default",
      "registerTime": 240.520375,
      "resourceTable": Object {
        "host": Array [
          null,
          null,
          null,
          null,
          null,
        ],
        "length": 5,
        "lib": Array [
          2,
          1,
          4,
          3,
          5,
        ],
        "name": Array [
          0,
          2,
          5,
          8,
          14,
        ],
        "type": Array [
          1,
          1,
          1,
          1,
          1,
        ],
      },
      "samples": Object {
        "length": 9,
        "stack": Array [
          5,
          5,
          10,
          10,
          15,
          15,
          20,
          20,
          24,
        ],
        "threadCPUDelta": Array [
          6,
          0,
          9,
          0,
          20,
          0,
          6,
          0,
          2,
        ],
        "time": Array [
          240.520375,
          251.722709,
          252.728334,
          261.724667,
          262.732625,
          273.755,
          274.768584,
          285.764042,
          286.765084,
        ],
        "weight": Array [
          1,
          11,
          1,
          9,
          1,
          11,
          1,
          11,
          1,
        ],
        "weightType": "samples",
      },
      "stackTable": Object {
        "frame": Array [
          0,
          1,
          2,
          3,
          4,
          5,
          6,
          2,
          3,
          4,
          5,
          6,
          2,
          3,
          4,
          5,
          6,
          2,
          3,
          4,
          5,
          7,
          8,
          9,
          10,
        ],
        "length": 25,
        "prefix": Array [
          null,
          0,
          1,
          2,
          3,
          4,
          1,
          6,
          7,
          8,
          9,
          6,
          11,
          12,
          13,
          14,
          11,
          16,
          17,
          18,
          19,
          null,
          21,
          22,
          23,
        ],
      },
      "stringArray": Array [
        "libsystem_pthread.dylib",
        "0x6f93",
        "a.out",
        "0x3db3",
        "0x3d33",
        "libsystem_c.dylib",
        "0xd47f",
        "0xd567",
        "libsystem_kernel.dylib",
        "0x43e8",
        "0x3d67",
        "0x6f9f",
        "0x769f",
        "0x4983",
        "libsystem_platform.dylib",
        "0x3f3c",
        "threadfunc(void*)",
        "fac(unsigned long)",
        "_pthread_start",
        "_pthread_exit",
        "_pthread_terminate_invoke",
        "__semwait_signal",
        "usleep",
        "nanosleep",
        "_platform_memset",
      ],
      "tid": "6274164",
      "unregisterTime": 287.756375,
    },
    Object {
      "frameTable": Object {
        "address": Array [
          28563,
          15795,
          15667,
          54399,
          54631,
          17384,
          15719,
        ],
        "category": Array [
          1,
          1,
          1,
          1,
          1,
          1,
          1,
        ],
        "column": Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "func": Array [
          0,
          1,
          2,
          3,
          4,
          5,
          2,
        ],
        "inlineDepth": Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        "innerWindowID": Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "length": 7,
        "line": Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "nativeSymbol": Array [
          2,
          0,
          1,
          4,
          5,
          3,
          1,
        ],
        "subcategory": Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
      },
      "funcTable": Object {
        "columnNumber": Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "fileName": Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "isJS": Array [
          false,
          false,
          false,
          false,
          false,
          false,
          false,
        ],
        "length": 7,
        "lineNumber": Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ],
        "name": Array [
          13,
          11,
          12,
          15,
          16,
          14,
          10,
        ],
        "relevantForJS": Array [
          false,
          false,
          false,
          false,
          false,
          false,
          false,
        ],
        "resource": Array [
          0,
          1,
          1,
          2,
          2,
          3,
          1,
        ],
      },
      "isMainThread": false,
      "markers": Object {
        "category": Array [],
        "data": Array [],
        "endTime": Array [],
        "length": 0,
        "name": Array [],
        "phase": Array [],
        "startTime": Array [],
      },
      "name": "Thread <6274165>",
      "nativeSymbols": Object {
        "address": Array [
          15752,
          15644,
          28428,
          17376,
          54332,
          54412,
        ],
        "functionSize": Array [
          76,
          108,
          320,
          44,
          80,
          464,
        ],
        "length": 6,
        "libIndex": Array [
          1,
          1,
          2,
          3,
          4,
          4,
        ],
        "name": Array [
          11,
          12,
          13,
          14,
          15,
          16,
        ],
      },
      "pausedRanges": Array [],
      "pid": "56127",
      "processName": "a.out",
      "processShutdownTime": 300.814417,
      "processStartupTime": 237.805292,
      "processType": "default",
      "registerTime": 240.520375,
      "resourceTable": Object {
        "host": Array [
          null,
          null,
          null,
          null,
        ],
        "length": 4,
        "lib": Array [
          2,
          1,
          4,
          3,
        ],
        "name": Array [
          0,
          2,
          5,
          8,
        ],
        "type": Array [
          1,
          1,
          1,
          1,
        ],
      },
      "samples": Object {
        "length": 10,
        "stack": Array [
          5,
          5,
          10,
          10,
          15,
          15,
          20,
          20,
          25,
          25,
        ],
        "threadCPUDelta": Array [
          2,
          0,
          6,
          0,
          4,
          0,
          5,
          0,
          4,
          0,
        ],
        "time": Array [
          240.520375,
          251.722709,
          252.728334,
          263.723459,
          264.733042,
          274.768584,
          275.7785,
          287.756375,
          288.76475,
          299.7795,
        ],
        "weight": Array [
          1,
          11,
          1,
          11,
          1,
          10,
          1,
          12,
          1,
          11,
        ],
        "weightType": "samples",
      },
      "stackTable": Object {
        "frame": Array [
          0,
          1,
          2,
          3,
          4,
          5,
          6,
          2,
          3,
          4,
          5,
          6,
          2,
          3,
          4,
          5,
          6,
          2,
          3,
          4,
          5,
          6,
          2,
          3,
          4,
          5,
        ],
        "length": 26,
        "prefix": Array [
          null,
          0,
          1,
          2,
          3,
          4,
          1,
          6,
          7,
          8,
          9,
          6,
          11,
          12,
          13,
          14,
          11,
          16,
          17,
          18,
          19,
          16,
          21,
          22,
          23,
          24,
        ],
      },
      "stringArray": Array [
        "libsystem_pthread.dylib",
        "0x6f93",
        "a.out",
        "0x3db3",
        "0x3d33",
        "libsystem_c.dylib",
        "0xd47f",
        "0xd567",
        "libsystem_kernel.dylib",
        "0x43e8",
        "0x3d67",
        "threadfunc(void*)",
        "fac(unsigned long)",
        "_pthread_start",
        "__semwait_signal",
        "usleep",
        "nanosleep",
      ],
      "tid": "6274165",
      "unregisterTime": 300.814417,
    },
  ],
}
`;
